In [1]:
from openai import OpenAI
import json, os, sys
import os
import re
import random

In [2]:
client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

In [3]:
def generate_difference(chunk, client, model="gpt-4o"):
    prompt = f"""
You are an expert in news analysis and skilled at capturing narrative difference. 
I have some news articles in English and Chinese and I need you to help me identify the narrative differences between them.

{chunk}


"""
    messages = [{"role": "user", "content": prompt}]
    try:
        response = client.chat.completions.create(
            model=model, messages=messages, temperature=0
        )

        content = response.choices[0].message.content
        return content
    except Exception as e:  # if the model fails to return a response
        print(f"Error: {e}")
        return "Sorry, error from GPT."

In [4]:
def process_batches(input_file, output_file, client, model, label_a_index):
    # Read the article from the input JSON file
    with open(input_file, "r") as f:
        articles = json.load(f)

    # Extract summaries from the articles
    article_body = [article["body"] for article in articles]

    # get the article body with the label_a_index list

    article_a = [article_body[i] for i in label_a_index]

    # convert the article to JSON string
    chunk = json.dumps(article_a)

    responses = []

    difference = generate_difference(chunk, client, model=model)
    responses.append(
        {
            "Difference": difference,
        }
    )

    # Write the generated questions to the output JSON file
    with open(output_file, "w") as f:
        json.dump(responses, f, indent=4)

In [22]:
import pandas as pd

# read the label file
labels = pd.read_csv("labels_v3.csv", header=None, names=["label"])

In [27]:
# get the label index if labe is 1
label_a_index = list(labels[labels["label"] == 1].index)
# label_b_index = list(labels[labels["label"] == 9].index)
random.shuffle(label_a_index)

In [28]:
label_a_index = label_a_index[:20]
label_a_index

[25, 29, 2, 96, 143, 91, 46, 113, 48, 3]

In [29]:
input_file = "combined_Russo_Ukrainian_War.json"
output_file = "output_difference_v3.json"

# Assuming you have your OpenAI API client initialized as `client`
process_batches(
    input_file,
    output_file,
    client,
    model="gpt-4o",
    label_a_index=label_a_index,
)

In [30]:
# print the value from JSON file

with open(output_file, "r") as f:
    output = json.load(f)[0]
    print(output["Difference"])

To identify the narrative differences between the English and Chinese news articles, let's break down the key points and themes in each language.

### English Articles:
1. **U.S. Military Aid to Ukraine**:
   - The U.S. Department of Defense announced a $150 million Presidential Drawdown Authority (PDA) package for Ukraine, including air defense interceptors, artillery, anti-tank weapons, and other military equipment.
   - An additional $2.2 billion in Ukraine Security Assistance Initiative (USAI) funds will be used to purchase interceptors for Patriot and NASAMS air defense systems.
   - The U.S. aims to accelerate the delivery of these munitions by re-sequencing upcoming Foreign Military Sales deliveries.
   - The U.S. continues to work with around 50 allies and partners to support Ukraine against Russian aggression.

2. **Disinformation and Propaganda**:
   - A false claim circulated on social media that Ukraine's First Lady bought a $4 million Bugatti sports car using aid money.
  

# Cluster 1

To identify the narrative differences between the English and Chinese news articles, let's break down the key points and themes in each language.

### English Articles:
1. **U.S. Military Aid to Ukraine**:
   - The U.S. Department of Defense announced a $150 million Presidential Drawdown Authority (PDA) package for Ukraine, including air defense interceptors, artillery, anti-tank weapons, and other military equipment.
   - An additional $2.2 billion in Ukraine Security Assistance Initiative (USAI) funds will be used to purchase interceptors for Patriot and NASAMS air defense systems.
   - The U.S. aims to accelerate the delivery of these munitions by re-sequencing upcoming Foreign Military Sales deliveries.
   - The U.S. continues to work with around 50 allies and partners to support Ukraine against Russian aggression.

2. **Disinformation and Propaganda**:
   - A false claim circulated on social media that Ukraine's First Lady bought a $4 million Bugatti sports car using aid money.
   - The claim was debunked by Bugatti and identified as Russian disinformation.
   - The disinformation was traced back to a French publication, Verite Cachee, which has a pro-Russian slant and was created shortly before the NATO summit.

3. **International Support**:
   - The Netherlands is providing F-16 jets and training for Ukrainian pilots.
   - Germany, despite providing military aid, has promised not to become a party to the conflict.
   - The U.S. and its allies continue to provide critical capabilities to Ukraine to counter Russian aggression.

### Chinese Articles:
1. **U.S. Military Aid to Ukraine**:
   - The U.S. Department of Defense announced a new $23 billion aid package for Ukraine, including key air defense interceptors and anti-tank weapons.
   - The announcement was made during a meeting between U.S. Defense Secretary Lloyd Austin and Ukrainian Defense Minister Rustem Umerov.
   - The U.S. has been a major military supporter of Ukraine since the conflict began, providing over $50 billion in military aid.

2. **Support for Ukraine's NATO Membership**:
   - During the meeting, Umerov expressed Ukraine's desire to join NATO and hoped for an invitation soon.
   - Austin supported Ukraine's NATO aspirations and mentioned that measures would be taken to build a bridge for Ukraine to become a NATO member.

3. **Concerns about Future U.S. Support**:
   - There are concerns about the future of U.S. support for Ukraine if Donald Trump wins the presidential election in November.
   - Trump's advisors have proposed a plan to end the Russia-Ukraine war, which would require Ukraine to engage in peace talks with Moscow to receive more U.S. weapons.

### Narrative Differences:
1. **Focus on Aid Amounts**:
   - The English articles emphasize the specific amounts and types of military aid being provided, focusing on the immediate and practical aspects of the support.
   - The Chinese articles mention a larger aid amount ($23 billion) and provide a broader context of U.S. support since the conflict began.

2. **Disinformation and Propaganda**:
   - The English articles highlight the issue of Russian disinformation, specifically debunking a false claim about Ukraine's First Lady.
   - The Chinese articles do not mention this aspect, focusing instead on the official military aid and political support.

3. **NATO Membership and Future Support**:
   - The Chinese articles discuss Ukraine's aspirations to join NATO and the support from the U.S. for this goal.
   - They also highlight concerns about the continuity of U.S. support depending on the outcome of the upcoming U.S. presidential election.
   - The English articles do not delve into Ukraine's NATO membership aspirations or the potential impact of the U.S. election on future support.

4. **Tone and Emphasis**:
   - The English articles maintain a tone of solidarity and urgency, emphasizing the collective effort of the U.S. and its allies to support Ukraine.
   - The Chinese articles have a more cautious tone, acknowledging the support but also highlighting uncertainties and potential political shifts in the U.S.

By analyzing these differences, we can see how the same events are framed differently in English and Chinese media, reflecting varying priorities and concerns in the narratives.

## Cluster 0

The English and Chinese articles present different narratives and focal points regarding the Ukraine conflict and related geopolitical dynamics. Here are the key narrative differences:

1. **Technological Innovation and Military Strategy**:
   - **English**: The English articles emphasize Ukraine's technological advancements, particularly in autonomous drones and AI-driven weaponry. They highlight Ukraine's transformation into a "Silicon Valley" for military tech, driven by necessity and innovation. The narrative focuses on the ethical and strategic implications of these technologies, including potential global arms races and the future of warfare.
   - **Chinese**: The Chinese articles focus more on the immediate military engagements and outcomes, such as the effectiveness of Ukrainian and Russian military strategies, the use of drones, and specific battles. There is less emphasis on technological innovation and more on the tactical and operational aspects of the conflict.

2. **Geopolitical Context and Alliances**:
   - **English**: The English articles provide a broader geopolitical context, discussing NATO's role, the implications of Ukraine's potential NATO membership, and the international response to the conflict. They also cover the strategic interests of the US and its allies, and the potential long-term impacts on global security.
   - **Chinese**: The Chinese articles tend to focus on the immediate geopolitical maneuvers, such as the involvement of North Korean soldiers, the impact of Western military aid, and the specific actions of countries like Germany and the US. There is a strong emphasis on the direct consequences of these actions on the battlefield.

3. **Humanitarian and Ethical Concerns**:
   - **English**: The English articles often touch on the humanitarian and ethical concerns related to the conflict, such as the use of autonomous weapons, the potential for civilian casualties, and the broader implications for international law and human rights.
   - **Chinese**: The Chinese articles are more focused on the practical and immediate outcomes of the conflict, such as the effectiveness of military strategies and the impact on the involved nations' military capabilities. There is less emphasis on the ethical and humanitarian dimensions.

4. **Narrative Tone and Perspective**:
   - **English**: The tone in the English articles is analytical and often speculative, considering future scenarios and the broader implications of current events. There is a focus on the strategic and technological evolution of warfare.
   - **Chinese**: The tone in the Chinese articles is more immediate and descriptive, focusing on the current state of the conflict, specific military engagements, and the direct actions of the involved parties. The narrative is more about reporting the ongoing developments rather than speculating on future implications.

5. **Focus on Specific Events and Figures**:
   - **English**: The English articles often highlight specific technological advancements, interviews with key figures, and detailed descriptions of new military technologies. They provide a more in-depth look at the innovations driving the conflict.
   - **Chinese**: The Chinese articles focus more on the broader military and geopolitical landscape, including the involvement of various countries, the impact of military aid, and the strategic decisions of key players. There is a greater emphasis on the collective actions and responses of nations rather than individual innovations.

In summary, the English articles provide a detailed, technology-focused, and speculative analysis of the Ukraine conflict, while the Chinese articles offer a more immediate, tactical, and descriptive account of the ongoing military and geopolitical developments.

## V2.1

The narratives in the English and Chinese articles about the Ukraine conflict reveal distinct perspectives and emphases. Here are the key differences:

1. **Focus on Human Experience vs. Military Analysis**:
   - **Chinese Articles**: The Chinese narrative often focuses on personal stories and the human experience of the conflict. For example, the detailed account of a Chinese student returning to Ukraine amidst the war highlights the daily struggles, such as power outages and the emotional toll of the conflict.
   - **English Articles**: The English articles, particularly those from UN reports and Western media, emphasize the broader humanitarian impact, casualty statistics, and the geopolitical implications of the conflict. They provide detailed accounts of civilian casualties, infrastructure damage, and international diplomatic efforts.

2. **Narrative Tone**:
   - **Chinese Articles**: The tone in the Chinese articles tends to be more personal and reflective, often using first-person narratives to convey the impact of the war on individuals. This approach creates a more intimate and emotional connection with the reader.
   - **English Articles**: The tone in the English articles is more formal and analytical, focusing on factual reporting and statistical data. The narrative is often framed within the context of international law, human rights, and geopolitical strategy.

3. **Emphasis on Military Developments**:
   - **Chinese Articles**: There is a significant focus on specific military engagements and the technical aspects of warfare, such as the capabilities of different aircraft and the strategic use of drones. The articles discuss the effectiveness of various military technologies and tactics in detail.
   - **English Articles**: While military developments are covered, the emphasis is more on the consequences of these actions, such as civilian casualties and the destruction of infrastructure. The articles also highlight the international response and the support provided to Ukraine by Western allies.

4. **Geopolitical Context**:
   - **Chinese Articles**: The geopolitical context in the Chinese articles often includes a broader discussion of international relations, including the roles of countries like the United States, Russia, and European nations. There is also mention of the potential impact on global security and economic stability.
   - **English Articles**: The geopolitical context is framed around the actions and statements of international organizations like the UN and NATO. There is a strong focus on the legal and moral implications of the conflict, as well as the efforts to hold Russia accountable for its actions.

5. **Humanitarian Impact**:
   - **Chinese Articles**: The humanitarian impact is conveyed through personal anecdotes and the challenges faced by individuals, such as students and local residents. The articles describe the day-to-day difficulties and the resilience of the people affected by the war.
   - **English Articles**: The humanitarian impact is presented through comprehensive reports and statistical data. The articles highlight the scale of the crisis, including the number of displaced persons, the extent of infrastructure damage, and the efforts of international aid organizations.

6. **Diplomatic Efforts**:
   - **Chinese Articles**: Diplomatic efforts are discussed in the context of broader international relations, with an emphasis on the positions and actions of major powers. There is also mention of the potential for peace talks and the conditions required for a ceasefire.
   - **English Articles**: Diplomatic efforts are detailed with specific references to meetings, agreements, and statements by international leaders and organizations. The articles often include direct quotes from officials and provide analysis of the potential outcomes of these diplomatic initiatives.

In summary, the Chinese articles provide a more personal and detailed account of the human and military aspects of the conflict, while the English articles focus on the broader humanitarian impact, geopolitical context, and diplomatic efforts. Both narratives offer valuable insights but cater to different aspects of the ongoing crisis.

## V2

The news articles in English and Chinese present different narratives and emphases regarding the Ukraine-Russia conflict. Here are the key narrative differences:

1. **Focus on Public Opinion and Polling**:
   - **English**: The English articles emphasize public opinion and polling data, particularly from the European Council on Foreign Relations (ECFR). They discuss sustained public support for military aid to Ukraine, divisions on how the war will end, and the purpose of European aid. The narrative highlights the challenges NATO leaders face in reconciling differing opinions between Europeans and Ukrainians.
   - **Chinese**: The Chinese articles do not focus on public opinion or polling data. Instead, they emphasize geopolitical maneuvers, military developments, and the involvement of various countries in the conflict.

2. **Geopolitical Analysis and International Relations**:
   - **English**: The English articles provide detailed analysis of international relations, including the roles of NATO, the EU, and the US. They discuss the implications of political shifts in Europe, the impact of US domestic politics on the conflict, and the strategic importance of maintaining unity within NATO.
   - **Chinese**: The Chinese articles focus more on the geopolitical implications of the conflict, including the involvement of countries like China, North Korea, and the dynamics within the EU. They highlight China's stance and its potential influence on the conflict, as well as the strategic moves by Russia and its allies.

3. **Military Developments and Aid**:
   - **English**: The English articles detail specific military aid packages from the US and other NATO countries, including the types of weapons and equipment being provided to Ukraine. They also discuss the strategic importance of these aid packages in bolstering Ukraine's defense capabilities.
   - **Chinese**: The Chinese articles provide a broader overview of military developments, including the involvement of North Korean troops and the strategic implications of military aid. They also discuss the potential impact of these developments on the overall conflict and the balance of power.

4. **Political Statements and Leadership**:
   - **English**: The English articles include statements from Western leaders, such as German Chancellor Olaf Scholz and former US President Donald Trump. They discuss the political rhetoric and promises made by these leaders regarding the conflict.
   - **Chinese**: The Chinese articles include statements from Chinese and Russian officials, as well as other international leaders. They focus on the diplomatic and political maneuvers being made by these countries and their leaders' perspectives on the conflict.

5. **Narrative Tone and Perspective**:
   - **English**: The English articles generally adopt a tone of cautious optimism about the potential for unity within NATO and the effectiveness of military aid. They emphasize the importance of continued support for Ukraine and the challenges of achieving a negotiated settlement.
   - **Chinese**: The Chinese articles adopt a more critical tone towards Western involvement in the conflict. They highlight the complexities and potential pitfalls of the conflict, including the risks of escalation and the strategic calculations of various countries.

6. **Specific Incidents and Developments**:
   - **English**: The English articles provide detailed accounts of specific incidents, such as the delay of US support packages and the political shifts in Europe. They also discuss the implications of these developments for the conflict.
   - **Chinese**: The Chinese articles provide a broader overview of recent developments, including military engagements, strategic moves by various countries, and the potential impact of these developments on the conflict.

In summary, the English articles focus more on public opinion, international relations, and specific military aid packages, while the Chinese articles emphasize geopolitical analysis, military developments, and the strategic implications of the conflict. The narratives reflect different perspectives and priorities, with the English articles highlighting the challenges and efforts to maintain unity within NATO, and the Chinese articles focusing on the broader geopolitical context and the involvement of various countries in the conflict.